rclcpp/include/rclcpp/timer.hpp

```diff
@@ -300,11 +318,26 @@ class GenericTimer : public TimerBase
     >::type * = nullptr
   >
   void
-  execute_callback_delegate()
+  execute_callback_delegate(const rcl_timer_call_info_t &)
   {
     callback_(*this);
   }
 
+
+  template<
+    typename CallbackT = FunctorT,
+    typename std::enable_if<
+      rclcpp::function_traits::same_arguments<CallbackT, TimerInfoCallbackType>::value
+    >::type * = nullptr
+  >
+  void
+  execute_callback_delegate(const rcl_timer_call_info_t & timer_call_info_)
+  {
+    const TimerInfo info{Time{timer_call_info_.expected_call_time, clock_->get_clock_type()},
+      Time{timer_call_info_.actual_call_time, clock_->get_clock_type()}};
+    callback_(info);
```

Comment:
We ran into the following problem :
System was running in Simulation mode, so it used the clock topic.
We have a timer processing odometry data at the same frequency as the clock update (100 Hz).

Also the Process was using the Multithreaded Executor. Therefore the reception of the next clock update is concurrent, as the timer and clock are in different callback groups.

On the start of each timer callback, we call node->now() to get the current time.
The problem we ran into was, that due to concurrency, It could happen, that the time jumped between the call to the timer callback and us calling node->now().
As a result of this, we got two times the same timestamp returned from node->now() in consecutive timer callbacks.

Reply:
Second nice thing, that occurred to me while writing the tests, you can detect timer overruns using this information.

Reply:
i see, thanks for the explanation. so especially simulation time, we can get to know the when the callback is actually fired with this PR.